pkg/apis/resource/validation/validation.go

```diff
@@ -433,179 +512,244 @@ func validateNodeName(name string, fldPath *field.Path) field.ErrorList {
+	allErrs = append(allErrs, validateResourceSliceSpec(&slice.Spec, nil, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateResourceSlice tests if a ResourceSlice update is valid.
+func ValidateResourceSliceUpdate(resourceSlice, oldResourceSlice *resource.ResourceSlice) field.ErrorList {
+	allErrs := corevalidation.ValidateObjectMetaUpdate(&resourceSlice.ObjectMeta, &oldResourceSlice.ObjectMeta, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateResourceSliceSpec(&resourceSlice.Spec, &oldResourceSlice.Spec, field.NewPath("spec"))...)
 	return allErrs
 }
 
-func validateResourceModel(model *resource.ResourceModel, fldPath *field.Path) field.ErrorList {
+func validateResourceSliceSpec(spec, oldSpec *resource.ResourceSliceSpec, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
-	entries := sets.New[string]()
-	if model.NamedResources != nil {
-		entries.Insert("namedResources")
-		allErrs = append(allErrs, namedresourcesvalidation.ValidateResources(model.NamedResources, fldPath.Child("namedResources"))...)
+	allErrs = append(allErrs, validateDriverName(spec.Driver, fldPath.Child("driver"))...)
+	allErrs = append(allErrs, validateResourcePool(spec.Pool, fldPath.Child("pool"))...)
+	nodeSelectionFields := sets.New[string]()
+	if spec.NodeName != "" {
+		nodeSelectionFields.Insert("nodeName")
+		allErrs = append(allErrs, validateNodeName(spec.NodeName, fldPath.Child("nodeName"))...)
 	}
-	switch len(entries) {
+	if spec.NodeSelector != nil {
+		nodeSelectionFields.Insert("nodeSelector")
+		allErrs = append(allErrs, corevalidation.ValidateNodeSelector(spec.NodeSelector, fldPath.Child("nodeSelector"))...)
+	}
+	if spec.AllNodes {
+		nodeSelectionFields.Insert("allNodes")
+	}
+	switch nodeSelectionFields.Len() {
 	case 0:
-		allErrs = append(allErrs, field.Required(fldPath, "exactly one structured model field must be set"))
+		allErrs = append(allErrs, field.Required(fldPath, "exactly one of nodeName, nodeSelector or allNodes is required"))
 	case 1:
-		// Okay.
 	default:
-		allErrs = append(allErrs, field.Invalid(fldPath, sets.List(entries), "exactly one field must be set, not several"))
+		allErrs = append(allErrs, field.Invalid(fldPath, sets.List(nodeSelectionFields), "these fields are mutually exclusive"))
+	}
+	allErrs = append(allErrs, validateSet(spec.Devices, resource.ResourceSliceMaxDevices, validateDevice,
+		func(device resource.Device) (string, string) {
+			return device.Name, "name"
+		}, fldPath.Child("devices"))...)
+	if oldSpec != nil {
+		allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(spec.Pool, oldSpec.Pool, fldPath.Child("pool"))...)
+		allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(spec.Driver, oldSpec.Driver, fldPath.Child("driver"))...)
+		allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(spec.NodeName, oldSpec.NodeName, fldPath.Child("nodeName"))...)
```

Comment:
The node name is special because it controls write access. It should never change.

For a slice with network-attached devices I can imagine that it may be useful to change the specification of where they are available.